### PR TITLE
0.13.0 PLAT 5120

### PR DESCRIPTION
**Pull request by lilachmaliniak**
Changes include Live and Web server side support for 404 reply, in case chunklist has expired.
Chunklist file is defined as expired if its age is older than configurable value.
Trigger to invalidate chunklist file is by the state machine of the EntryId, changing state to stopped. In that case, the chunklist files of all entry’s flavours are invalidated.
Still in progress, adding ENDLIST tag when no DC (media server) is available for the EntryId.

### CODE REVIEW

File: dvr-ws/routes/index.js [hunk 1]
```diff
@@ -2,18 +2,35 @@ var express = require('express');
...
     var fileName = req.params[1];
 
-    var disk = path.join(persistenceFormat.getEntryFullPath(entryId),fileName);
-    res.sendFile(disk);
+    var fullPath = path.join(persistenceFormat.getEntryFullPath(entryId),fileName);
+
+    if ( fileName.search('chunklist.m3u8') > -1 ) {
+
+        // Todo: take the configuration initialization outside
+        var chunklistExpireAge = config.get("webServerParams:chunklistExpireAge");
```
Comment (kalturaguy) [inline dvr-ws/routes/index.js]: please move it to global

File: dvr-ws/routes/index.js [hunk 2]
```diff
@@ -2,18 +2,35 @@ var express = require('express');
...
+
+        fsUtils.checkIsFileExpired(fullPath, chunklistExpireAge)
```
Comment (kalturaguy) [inline dvr-ws/routes/index.js]: checkIfFileExpired

**Comment by kaltura-hooks** [comment]
Hi @lilachmaliniak,
Thank you for contributing this pull request!
Please sign the Kaltura CLA so we can review and merge your contribution.
Learn more at http://bit.ly/KalturaContrib

File: dvr-ws/routes/index.js [hunk 3]
```diff
@@ -2,18 +2,35 @@ var express = require('express');
...
+            .then(function () {
+                res.sendFile(fullPath);
+            })
+            .catch(function () {
+                var message = fileName + '%s not found';
+                res.status(404).send(message);
```
Comment (kalturaguy) [inline dvr-ws/routes/index.js]: do we need content?  look what wowza response and akamai as well to figure out the best

File: lib/utils/fs-utils.js [hunk 4]
```diff
@@ -48,10 +50,65 @@ module.exports = (function(){
...
+
+        try {
+            fs.utimes(fullPath, fileModifiedTime, fileModifiedTime, function (err) {
+                logger.error('error: %s, failed to update modified time of %s to %s', err, that.fullPath, that.fileModifiedTime.toDateString());
+                res = false;
+            });
+
+        } catch (e) {
+            e = (e === undefined) ? new Error('suspecting undifined arg') : e;
+            logger.error('exception, failed to set modified time of %s to %s. error: %s', fullPath, fileModifiedTime.toDateString(), errorUtils.error2string(e));
```
Comment (kalturaguy) [inline lib/utils/fs-utils.js]: errorUtils.error2string already covers if it's null/undefined

File: lib/utils/fs-utils.js [hunk 5]
```diff
@@ -48,10 +50,65 @@ module.exports = (function(){
...
+                        resolve();
+                    else {
+                        logger.warn('modified time of %s is %s. File expired.', that.fullPath, stats.mtime.toDateString());
+                        reject();
+                    }
+
+                });
+
+            } catch (e) {
+                e = (e === undefined) ? new Error('internal error while trying to read file') : e;
```
Comment (kalturaguy) [inline lib/utils/fs-utils.js]: errorUtils.error2string already covers if it's null/undefined
 Add a line note

File: dvr-ws/routes/index.js (hunk 1 above)
Comment (lilachmaliniak) [inline dvr-ws/routes/index.js]: moved to global scope.

File: dvr-ws/routes/index.js (hunk 2 above)
Comment (lilachmaliniak) [inline dvr-ws/routes/index.js]: done

File: dvr-ws/routes/index.js (hunk 3 above)
Comment (lilachmaliniak) [inline dvr-ws/routes/index.js]: I agree that we don't so I removed the message and left the 404 status.

File: lib/utils/fs-utils.js (hunk 4 above)
Comment (lilachmaliniak) [inline lib/utils/fs-utils.js]: replaced line 65 and similar with calls to errorUtils.error2string.

File: lib/utils/fs-utils.js (hunk 5 above)
Comment (lilachmaliniak) [inline lib/utils/fs-utils.js]: replaced 97 to usage of errorUtils.error2string in log message.

**Comment by lilachmaliniak** [comment]
thanks for review comments :)

**Pull request closed without merging**